### PR TITLE
rbd: update ControllerExpandVolume with migration secret support

### PR DESCRIPTION
**Pull request by humblec**
This PR address below items from the tracker/todo: 
- [x] e2e: adjust migration e2e tests and introduce helper functions
    This commit adjust existing migration e2e tests to a couple of tests
    to cover the scenarios. The seperate filesystem and block tests have
    been shrinked to single one and also introduced a couple of helper
    functions to setup and teardown migraition specific secret,configmap
    and sc. The static pv function has been renamed to a general name
    while the tests were adjusted.
- [x]  rbd: add migration secret support to controllerserver functions
    This commit adds the migration secret request validation to expand,
    create controller functions.
- [x]  e2e: remove unparam linter escapes from getStaticPV and getStaticPVC
     This `unparam` linter escape is no longer needed and CI is failing
    if we keep there. This commit remove the same and make CI happy.
       
Updates #2509 

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by humblec** [comment]
The tests are failing due to the encryption vault failure.  However this PR specific tests have passed.  Appreciate review .. Thanks

File: internal/rbd/controllerserver.go
Comment (Madhu-1) [inline internal/rbd/controllerserver.go]: why this change is needed?

Comment (humblec) [reply internal/rbd/controllerserver.go]: this is required when we added migration secret support https://github.com/ceph/ceph-csi/pull/2562 , otherwise validation of the request will fail. @Madhu-1

Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: why you are adding it to createVolume? the migration is only to support the existing volume, isn't it?

Comment (humblec) [reply internal/rbd/controllerserver.go]: I am not adding anything here @Madhu-1, not sure what is the confusion here. Migration is for existing intree storage classes. As pointed out in all the migration documentations it is the very basic thing expected from migration,  for ex: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md
```
Dynamically Provisioned volumes will continue to be provisioned with the in-tree PersistentVolumeSource. The CSI external-provisioner to pick up the in-tree PVC's when migration is turned on and provision using the CSI Drivers;...
```

Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: Do you mean to say that the user will still create the CSI volumes using storageclass which contains the intree details?

Comment (humblec) [reply internal/rbd/controllerserver.go]: it is upto the admin , however mostly what I heard was that, new CSI SCs will be given to the users thus new volumes will be on CSI SC as anyway they are in migration path.

Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: IMHO we should not support this one in cephcsi and just stick to the plan to treat migration as static only add documentation to use new SC for csi volumes.

Comment (humblec) [reply internal/rbd/controllerserver.go]: This is very well understood and clear that new volumes has to go through and its a must requirement. We were always in plan and  never deviated from it . Also, this was  discussed in  CSI calls. `Static` treatment is only applicable to the existing intree volumes.

Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: @humblec can you please share with me the link to a  chat/doc where this is discussed in csi call? at least am not aware of any discussion. for migration, we are getting everything in bit and piece and this was not part of your design doc. now, this is news for me, and am not aware of it.

Comment (humblec) [reply internal/rbd/controllerserver.go]: >for migration, we are getting everything in bit and piece and this was not part of your design doc. now, this is news for me, and am not aware of it.
>
@Madhu-1 it was discussed  in csi stand up calls, may be you were unavailable on those calls. It was/is part of the tracker issue, design doc ( with above link in place)..etc and there was also an explicit  mention of  provisioning and other operations in the design doc which was removed when you asked to replace that section by this comment https://github.com/ceph/ceph-csi/pull/2609/files#r744505983 . That section was replaced by the review comment  ask as it was repeating the same thing.. It is really surprising that,  it is a news . Also, this is not breaking change at all!! All existing functionality works and helping users if they opted for migration.

Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: @nixpanic are you aware of this discussion?

Comment (humblec) [reply internal/rbd/controllerserver.go]: @Madhu-1 *at least* I remember discussing these as parf of these dates `2021-09-13 , 2021-09-14` in CSI standup calls. https://docs.google.com/document/d/1K1aerdMpraIh56-skdoEoVF9RZrO4NUcbHtjN-f3u1s/edit#heading=h.z6gjxa34z215 .  Atleast @Rakshith-R should be aware of this..Also I have provided updates in other stand up calls too. More or less, I still didnt get what is the `real concern` here? it is not breaking any existing functionality and rather helping users !!!

Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: @humblec at least I was not aware of any of these till I review this PR. the reason for asking to remove the duplicate content from the design doc here is that https://github.com/ceph/ceph-csi/pull/2609/files#r744505983 does not want to have the same exact copy-paste from the kubernetes design doc in cephcsi. if we get the design PR in first and then the implementation we can avoid these discussions and also we are doing multiple PR to achieve this one we can avoid doing cleanup introduced in the last PR like we are doing in this PR hope you got my concern

File: e2e/migration.go
Comment (Madhu-1) [inline e2e/migration.go:122]: is this a hack for e2e? because i don't believe the intree SC carry these parameters

File: internal/rbd/controllerserver.go
Comment (Rakshith-R) [reply internal/rbd/controllerserver.go:288]: @Madhu-1 *at least* I remember discussing these as parf of these dates `2021-09-13 , 2021-09-14` in CSI standup calls. https://docs.google.com/document/d/1K1aerdMpraIh56-skdoEoVF9RZrO4NUcbHtjN-f3u1s/edit#heading=h.z6gjxa34z215 .  Atleast @Rakshith-R should be aware of this.

I don't remember discussing create or resize of volume using intree storage class at that point in time.

>.Also I have provided updates in other stand up calls too. More or less, I still didnt get what is the `real concern` here? it is not breaking any existing functionality and rather helping users !!! 


IMO, We should be encouraging the usage of csi storageclass.

Comment (Rakshith-R) [reply internal/rbd/controllerserver.go]: >@Madhu-1 *at least* I remember discussing these as parf of these dates `2021-09-13 , 2021-09-14` in CSI standup calls. https://docs.google.com/document/d/1K1aerdMpraIh56-skdoEoVF9RZrO4NUcbHtjN-f3u1s/edit#heading=h.z6gjxa34z215 .  Atleast @Rakshith-R should be aware of this.
> 

 I don't remember discussing create or resize of volume using intree storage class at that point in time.
 
> >.Also I have provided updates in other stand up calls too. More or less, I still didnt get what is the `real concern` here? it is not breaking any existing functionality and rather helping users !!! 
> 



 IMO, We should be encouraging the usage of csi storageclass.
> 
> 
> 
>  
> 
>

File: e2e/rbd.go
Comment (Madhu-1) [inline e2e/rbd.go]: do we have Block PVC support for rbd in intree?

File: e2e/rbd.go
Comment (Madhu-1) [inline e2e/rbd.go]: remove with error

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: >@humblec at least I was not aware of any of these till I review this PR. the reason for asking to remove the duplicate content from the design doc here is that https://github.com/ceph/ceph-csi/pull/2609/files#r744505983 does not want to have the same exact copy-paste from the kubernetes design doc in cephcsi. if we get the design PR in first and then the implementation we can avoid these discussions and also we are doing multiple PR to achieve this one we can avoid doing cleanup introduced in the last PR like we are doing in this PR hope you got my concern
>

@Madhu-1 nw at all, I am glad that you asked the clarification. once again I was also surprised by reading this review comment. We can definitely make the process better if there are flaws , end of the day its all about helping users with the project.  :+1:

**Review by Madhu-1 — DISMISSED** [review]
a couple of things here

* cant we wait for kubernetes to release with the translation library? the whole e2e is a hack to test the migration not sure it works the way that is expected from the migration side. 
* I really don't see any value in supporting the dynamic provisioning for the intree storageclass. IMO we should document that for new dynamic PVC use new csi storageclass.\

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: >I don't remember discussing create or resize of volume using intree storage class at that point in time.' .

@Rakshith-R please refresh :), resize came in late, but new volumes and annotation hack was discussed :).

File: e2e/rbd.go
Comment (humblec) [reply e2e/rbd.go]: not neeed, I can remove that.

File: internal/rbd/controllerserver.go
Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: @humblec IMHO design discussion should include the maintainers :)

File: e2e/rbd.go
Comment (humblec) [reply e2e/rbd.go]: sure.

File: e2e/migration.go
Comment (humblec) [reply e2e/migration.go:122]: its a hack for e2e to simulate the situation of migration SC.

File: internal/rbd/controllerserver.go
Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: IMO the whole process is a complete mess for me at least. if we don't have proper path things become hard at least to understand the goal and to make decisions.

**Comment by humblec** [comment]
>   cant we wait for kubernetes to release with the translation library? the whole e2e is a hack to test the migration not sure it works the way that is expected from the migration side.
>

@Madhu-1   I added the e2e as it was the request since the very first PR of this effort. Indeed its good to have till we have 1.23 release is out. Once its out I will remove it.. 

>  I really don't see any value in supporting the dynamic provisioning for the intree storageclass. IMO we should document that for new dynamic PVC use new csi storageclass.\
>

Sure, we can document CSI SCs are recommend for new volumes in the doc. However we dont have a legitimate justification to block provisioning if the users really wanted to use intree SC. Also, I am afraid that, this is not an option at this stage as its the very basic requirement from kubernetes which is very well documented/set the expectations on since its start, also the change is very minor and not effecting/breaking any "existing" functionality at all !!    Also, as mentioned earlier, admin can control which SCs to submit to users...etc in a migration enabled cluster and its upto him and based on his requirement. Also it is a common thought process for all migration enabled plugin setups.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: @Madhu-1 indeed the process for feature enhancement can be improved.. I struggled on few topics too in the past. I can take an action item to have a discussion among the team/community here about the process and document the process here in the repo from next release onwards.. :+1:

File: e2e/rbd.go
Comment (humblec) [reply e2e/rbd.go]: done.

File: e2e/rbd.go
Comment (humblec) [reply e2e/rbd.go]: done.

File: internal/rbd/controllerserver.go
Comment (Rakshith-R) [reply internal/rbd/controllerserver.go]: > .@Rakshith-R please refresh :), resize came in late, but new volumes and annotation hack was discussed :).

@humblec  As far as i remember we mainly discussed mapping required for vol-id changes. 

I only supported the node operations for older intree pvc as they are crucial,
I still don't feel we should support other functions and instead should encourage users to switch to **csi storageclass** going forward. :)

Comment (humblec) [reply internal/rbd/controllerserver.go]: > > .@Rakshith-R please refresh :), resize came in late, but new volumes and annotation hack was discussed :).
> 
> @humblec As far as i remember we mainly discussed mapping required for vol-id changes.
> 
the vol id part  was for existing volumes, in that call new vols and annotations to identify or distinguish the volid were also discussed. any way there is no point of going back :)

>I only supported the node operations for older intree pvc as they are crucial,
>
for older intree PVCs we are only supporting node operations and then delete volume we are able to support.. creation is not for existing volumes, it is for existing SC.

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go]: Because this code is duplicated in multiple places now, you should use a helper function for it. Maybe introduce
```
cr, err := util.NewUserCredentialsWithMigration(req.GetSecrets())
```
as a variation on `util.NewUserCredentials()`.

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go]: This is quite ugly, the `CreateVolumeRequest` already contains secrets. Preventing to pass extra (or replacement) `secrets` would be much better. If there is really no other reasonable way to do this, it needs to be well documented for the function.

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go]: You can now just assign `err` on the previous line. It reduces the length of this line and makes it more readable.

File: internal/rbd/controllerserver.go
Comment (nixpanic) [reply internal/rbd/controllerserver.go]: The design doc should be updated to include what changes are specific for Ceph-CSI. That includes mentioning functions like `ParseAndSetSecretMapFromMigSecret()`, and when that needs to be used.

For migration, I was expecting that the old in-tree PVs would be migrated to new CSI PVs. But it seems there is no real migration of that  kind. It currently looks like in-tree PVs are not _migrated_, but rather support for the in-tree format is added to Ceph-CSI. This is a little confusing...

The design doc should probably describe in a little more detail what CSI procedures need support for the old in-tree PVs for _migration_, and how that functionality is added to Ceph-CSI. A separate PR for the (updated) design doc might be appropriate, and that should be approved/merged before implementing the needed changes (ideally).

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: @nixpanic sure, let me check , considering the secret parsing and using done in many places ( I mean here and there, ie sometimes from helper directly on request, sometimes direct usage of passed cred ..etc) I noticed some difficulty.. but let me recheck again.

**Comment by humblec** [comment]
>The design doc should be updated to include what changes are specific for Ceph-CSI. That includes mentioning functions like ParseAndSetSecretMapFromMigSecret(), and when that needs to be used.
>

Yeah, will add more details..

>For migration, I was expecting that the old in-tree PVs would be migrated to new CSI PVs. But it seems there is no real migration of that kind. It currently looks like in-tree PVs are not migrated, but rather support for the in-tree format is added to Ceph-CSI. This is a little confusing...

Thats correct .

>The design doc should probably describe in a little more detail what CSI procedures need support for the old in-tree PVs for migration, and how that functionality is added to Ceph-CSI. A separate PR for the (updated) design doc might be appropriate, and that should be approved/merged before implementing the needed changes (ideally).
>

I will add details to the doc. Thanks :+1:

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: > If there is really no other reasonable way to do this, it needs to be well documented for the function.
>
as we decided to not change the request itself,  this is the reasonable way to do this. I have also documented parseVolCreateRequest() with the details.. thanks.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: Sure.. done .. Thanks.

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go]: These `secrets` are only needed for KMS (encryption passphrase in the linked StorageClass Secret), so the migration secrets (cluster access credentials) are not needed.

```go
	secrets := req.GetSecrets()
	if util.IsMigrationSecret(secrets) {
		secrets, err = util.ParseAndSetSecretMapFromMigSecret(secrets)
		if err != nil {
			return nil, status.Error(codes.InvalidArgument, err.Error())
		}
	}

	cr, err := util.NewUserCredentials(secrets)
```

can therefore be replaced by a single  helper like

```go
cr, err := util.NewUserCredentialsWithMigration(req.GetSecrets());
```

or something along those lines. When the helper does not detect `IsMigrationSecret()`, it can return `NewUserCredentials()` directly. Such a single helper is much cleaner and easier to maintain copy/pasted code in different locations. It also reduces the :spaghetti: code sprinkled over multiple places for migration support.

Comment (humblec) [reply internal/rbd/controllerserver.go]: Sure, @nixpanic , will adopt that model here.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: converted to above model and all tests are passing :+1:  , ptal .. thanks @nixpanic

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: Done.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: Added https://github.com/ceph/ceph-csi/pull/2679 ..

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go]: what does `this model has been kept` mean in this comment?

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go:1445]: empty lines  between different sections make the code more readable, please do not remove them

File: internal/rbd/controllerserver.go
Comment (nixpanic) [inline internal/rbd/controllerserver.go:1455]: same here, keep empty lines between different code-blocks please

File: internal/util/credentials.go
Comment (nixpanic) [inline internal/util/credentials.go:145]: update the comment with the new function name too

File: internal/util/credentials.go
Comment (nixpanic) [inline internal/util/credentials.go]: that is `passedSecretMap`? Do you mean 'passed' as in the argument to the function? That seems rather redundant, and just `secrets` works fine everywhere else too.

Comment (humblec) [reply internal/util/credentials.go]: done.

File: internal/util/credentials.go
Comment (humblec) [reply internal/util/credentials.go:145]: Done.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go:1455]: Done.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go:1445]: Done.

File: internal/rbd/controllerserver.go
Comment (humblec) [reply internal/rbd/controllerserver.go]: forgot to remove this comment, it made sense only with previous version, no longer needed above explanation. Cleaned up now.. Thanks

**Comment by humblec** [comment]
@nixpanic all comments are addressed, ptal. Thanks :+1:

File: internal/util/credentials.go
Comment (nixpanic) [inline internal/util/credentials.go:161]: this is a new helper function, so there should probably be a unit test for that?

**Review by nixpanic — COMMENTED** [review]
Looks ok to me. There were lots of discussions with @Rakshith-R and @Madhu-1 , so I would like to see their approvals too.

File: internal/util/credentials.go
Comment (humblec) [reply internal/util/credentials.go:161]: actually this function make use of 2 other helpers which already covered in the unit test. The part which is not covered with the unit test is for the core function `newUserWithCredentials` which existed for some time , but we havent added any unit tests because of its nature like creation of keyfile inside the tmp dir. so I think may be unit test is not required for this new helper function..

**Comment by humblec** [comment]
> Looks ok to me. There were lots of discussions with @Rakshith-R and @Madhu-1 , so I would like to see their approvals too.

Thanks @nixpanic :+1: 

@Madhu-1 @Rakshith-R can you please review this PR for the latest version? I have also added more details to the doc PR.

**Review by Madhu-1 — COMMENTED** [review]
Kubernetes 1.23 is planned for Dec 7, IMO we can remove all the e2e hacks we don't need to rework again after some time and do real e2e tests to make sure everything is working fine?

**Comment by humblec** [comment]
> Kubernetes 1.23 is planned for Dec 7, IMO we can remove all the e2e hacks we don't need to rework again after some time and do real e2e tests to make sure everything is working fine?

@Madhu-1 this PR is not introducing any extra e2e tests but  rather it reduced the e2e tests from 3 to 2.  Considering this PR has already passed 2 weeks here , can we get this in ? also no more hack is required on e2e further. Once kube release is available we can consume but imo we dont have to wait for that.

**Comment by Madhu-1** [comment]
> > Kubernetes 1.23 is planned for Dec 7, IMO we can remove all the e2e hacks we don't need to rework again after some time and do real e2e tests to make sure everything is working fine?
> 
> @Madhu-1 this PR is not introducing any extra e2e tests but rather it reduced the e2e tests from 3 to 2. Considering this PR has already passed 2 weeks here , can we get this in ? also no more hack is reduced on e2e further. Once kube release is available we can consume but imo we dont have to wait for that.

@nixpanic  what's your thought on this one?

**Review by nixpanic — APPROVED** [review]

**Comment by humblec** [comment]
Thanks a lot @nixpanic :+1:

**Comment by humblec** [comment]
@Madhu-1 can you please review/approve this , I would like to wrap this without further delay..

**Comment by Madhu-1** [comment]
@humblec  can you please don't dismiss the review requests. if any changes are requested and someone else comes and approves the PR it will get merged.

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**
